File: static/game.js
// ── State ────────────────────────────────────────────────────────────────────
let state = null;
let selectedCard = null;
let handRevealed = false;
let mode = 'human';                 // 'human' (hot-seat) | 'greedy' | 'probabilistic'
let beliefVisible = false;          // belief histogram toggle (probabilistic bot only)
const isBot = () => mode !== 'human';
const hasBelief = () => mode === 'probabilistic';

// ── API ──────────────────────────────────────────────────────────────────────
let busy = false;                   // a request is in flight — block new actions

async function api(endpoint, method = 'GET', body = null) {
    const opts = { method, headers: { 'Content-Type': 'application/json' } };
    if (body !== null) opts.body = JSON.stringify(body);
    const res = await fetch(`/api/${endpoint}`, opts);
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    return res.json();
}

// Run a network action with a single-flight lock. Blocks concurrent clicks
// (e.g. while the bot is thinking) and, on any failure, re-syncs the client
// from the server so the UI can never get wedged out of phase.
async function act(fn) {
    if (busy) return;
    busy = true;
    setBusyUI(true);
    try {
        await fn();
    } catch (e) {
        try { state = await api('state'); } catch (_) {}
        showMsg('Connection hiccup — resynced with the table.', true);
    } finally {
        busy = false;
        setBusyUI(false);
        render();
    }
}

function setBusyUI(on) {
    document.body.classList.toggle('busy', on);
    if (on) el('message').textContent = 'Working…';
    updateButtons();
}

// ── Actions ──────────────────────────────────────────────────────────────────
async function newGame() {
    mode = el('opponent-select').value;
    await act(async () => {
        state = await api('new_game', 'POST', { opponent: mode });
        selectedCard = null;
        handRevealed = isBot();      // bot mode: human's view is always up, no pass screen
    });
}

async function handleStockClick() {
    if (busy || !state || state.phase !== 'draw' || state.game_over || !handRevealed) return;
    await act(async () => {
        state = await api('draw', 'POST', { source: 'stock' });
        selectedCard = null;
    });
}

async function handleDiscardClick() {
    if (busy || !state || state.phase !== 'draw' || state.game_over || !handRevealed) return;
    if (!state.discard_top) return;
    await act(async () => {
        state = await api('draw', 'POST', { source: 'discard' });
        selectedCard = null;
    });
}

async function doDiscard() {
    if (busy || !selectedCard) return;
    await act(async () => {
        const resp = await api('discard', 'POST', { card: selectedCard });
        if (!resp.ok) { showMsg(resp.error, true); return; }
        state = resp;
        selectedCard = null;
        if (!state.game_over) handRevealed = isBot();   // hot-seat hides for the pass; bot stays revealed
    });
}

async function doKnock() {
    if (busy || !selectedCard) return;
    await act(async () => {
        const resp = await api('knock', 'POST', { card: selectedCard });
        if (!resp.ok) { showMsg(resp.error, true); return; }
        state = resp;
        selectedCard = null;
    });
}

function revealHand() {
    handRevealed = true;
    render();
}

function toggleBelief() {
    beliefVisible = !beliefVisible;
    render();
}

// ── Card helpers ─────────────────────────────────────────────────────────────
const SUIT_SYM = { H: '♥', D: '♦', C: '♣', S: '♠' };
const RED = new Set(['H', 'D']);

function parseCard(str) {
    return str.startsWith('10') ? ['10', str[2]] : [str[0], str[1]];
}

function makeCard(cardStr, { faceDown = false, selected = false, inMeld = false,
    deadwood = false, drawn = false, clickable = false, noHover = false } = {}) {

    const el = document.createElement('div');
    const cls = ['card'];
    if (faceDown) cls.push('face-down');
    if (selected) cls.push('selected');
    if (inMeld) cls.push('in-meld');
    if (deadwood) cls.push('deadwood-card');
    if (drawn && !selected) cls.push('drawn');
    if (noHover) cls.push('no-hover');
    el.className = cls.join(' ');

    if (faceDown) return el;

    const [rank, suit] = parseCard(cardStr);
    if (RED.has(suit)) el.classList.add('red');
    const sym = SUIT_SYM[suit];

    el.innerHTML = `
        <div class="c-tl"><span class="c-rank">${rank}</span><span class="c-suit-sm">${sym}</span></div>
        <span class="c-suit-lg">${sym}</span>
        <div class="c-br"><span class="c-rank">${rank}</span><span class="c-suit-sm">${sym}</span></div>
    `;

    if (clickable) el.addEventListener('click', () => onCardClick(cardStr));
    return el;
}

// ── Interaction ───────────────────────────────────────────────────────────────
function onCardClick(cardStr) {
    if (busy || !state || state.phase !== 'discard' || state.game_over || !handRevealed) return;
    selectedCard = selectedCard === cardStr ? null : cardStr;
    renderCurrentHand();
    updateButtons();
    updateDeadwoodHint();
}

// ── Render ───────────────────────────────────────────────────────────────────
function render() {
    if (!state) return;
    updateBeliefButton();

    if (state.game_over) {
        hide('pass-overlay');
        showGameOver();
        return;
    }

    hide('gameover-overlay');

    if (!handRevealed && !isBot()) {
        showPassScreen();
        return;
    }

    hide('pass-overlay');
    renderGame();
}

function renderGame() {
    const p = state.current_player;
    const opp = 1 - p;

    // Scores
    el('score-p1').textContent = state.scores[0];
    el('score-p2').textContent = state.scores[1];

    // Labels
    el('label-current').textContent = isBot() ? 'You' : `Player ${p + 1}`;
    el('label-current').className = 'player-label active';
    el('label-opponent').textContent = isBot()
        ? (state.opponent_name || 'Bot')
        : `Player ${opp + 1}`;
    el('label-opponent').className = 'player-label';

    // Opponent hand (face down)
    const oppHandEl = el('hand-opponent');
    oppHandEl.innerHTML = '';
    state.hands[opp].forEach(() => oppHandEl.appendChild(makeCard(null, { faceDown: true })));

    // Stock
    el('stock-count').textContent = state.stock_count;
    el('stock-pile').className = 'pile' + (state.phase === 'draw' ? ' can-draw' : '');

    // Discard top
    rebuildDiscardTop();
    el('discard-pile').className = 'pile' + (state.phase === 'draw' && state.discard_top ? ' can-draw' : '');

    // Current hand + UI
    renderCurrentHand();
    updateButtons();
    updateDeadwoodHint();
    renderBelief();
    showMsg(state.message);
}

function renderCurrentHand() {
    const p = state.current_player;
    const handEl = el('hand-current');
    handEl.innerHTML = '';

    // Own-melds view: server splits the active hand into melds vs deadwood.
    const view = state.own_view || {};
    const meldCards = new Set((view.current_melds || []).flat());

    state.hands[p].forEach(c => {
        handEl.appendChild(makeCard(c, {
            selected: c === selectedCard,
            drawn: c === state.drawn_card,
            inMeld: meldCards.has(c),
            deadwood: meldCards.size > 0 && !meldCards.has(c),
            clickable: true,
        }));
    });
}

// ── Belief histogram (probabilistic bot) ───────────────────────────────────────
const RANKS = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'];
const SUITS = ['S', 'H', 'D', 'C'];

function updateBeliefButton() {
    const btn = el('btn-belief');
    btn.style.display = hasBelief() ? '' : 'none';
    btn.textContent = beliefVisible ? 'Hide belief' : 'Show belief';
}

function renderBelief() {
    updateBeliefButton();
    const panel = el('belief-panel');
    if (!hasBelief() || !beliefVisible || !state || !state.belief) {
        hide('belief-panel');
        return;
    }
    show('belief-panel');

    const sum = state.belief_sum;
    el('belief-sum').textContent = sum != null ? `Σ = ${sum.toFixed(2)}` : '';

    const grid = el('belief-grid');
    grid.innerHTML = '';

    // Header row: blank corner + rank labels
    grid.appendChild(beliefLabel(''));
    RANKS.forEach(r => grid.appendChild(beliefLabel(r)));

    SUITS.forEach(s => {
        grid.appendChild(beliefLabel(SUIT_SYM[s], RED.has(s)));
        RANKS.forEach(r => {
            const cardStr = `${r}${s}`;
            const p = state.belief[cardStr] ?? 0;
            grid.appendChild(beliefCell(p));
        });
    });
}

function beliefLabel(text, red = false) {
    const d = document.createElement('div');
    d.className = 'belief-axis' + (red ? ' red' : '');
    d.textContent = text;
    return d;
}

function beliefCell(p) {
    const d = document.createElement('div');
    d.className = 'belief-cell';
    if (p >= 0.999) d.classList.add('certain');       // known held (P=1)
    else if (p <= 0.001) d.classList.add('dead');     // impossible (P=0)
    // accent fill scaled by probability
    d.style.background = `rgba(0, 170, 204, ${Math.min(1, p).toFixed(3)})`;
    d.textContent = p >= 0.005 ? p.toFixed(2).slice(1) : '';  // ".42" style, blank if ~0
    d.title = p.toFixed(3);
    return d;
}

function rebuildDiscardTop() {
    const old = el('discard-top');
    let card;
    if (state.discard_top) {
        card = makeCard(state.discard_top, { noHover: true });
    } else {
        card = document.createElement('div');
        card.className = 'card placeholder';
        card.innerHTML = '<span class="placeholder-text">—</span>';
    }
    card.id = 'discard-top';
    old.replaceWith(card);
}

function updateButtons() {
    const canAct = !busy && state && state.phase === 'discard' && selectedCard !== null && handRevealed;
    el('btn-discard').disabled = !canAct;
    el('btn-knock').disabled = !canAct;
}

function updateDeadwoodHint() {
    const hint = el('deadwood-hint');
    if (!state || !handRevealed || !state.own_view) {
        hint.textContent = '';
        return;
    }
    const view = state.own_view;
    const dw = view.current_deadwood_value;
    const meldCount = (view.current_melds || []).length;
    const meldTxt = meldCount === 1 ? '1 meld' : `${meldCount} melds`;
    let txt = `Deadwood ${dw} · ${meldTxt}`;
    if (state.phase === 'discard') {
        txt += dw <= 10 ? ' — you can knock' : ' — knock needs deadwood ≤ 10';
    }
    hint.textContent = txt;
}

// ── Pass screen ───────────────────────────────────────────────────────────────
function showPassScreen() {
    const p = state.current_player;
    el('pass-title').textContent = `Player ${p + 1}'s Turn`;
    el('pass-player').textContent = `Player ${p + 1}`;
    show('pass-overlay');
    hide('gameover-overlay');
}

// ── Game over screen ──────────────────────────────────────────────────────────
function showGameOver() {
    el('score-p1').textContent = state.scores[0];
    el('score-p2').textContent = state.scores[1];

    const ki = state.knock_info;
    let title = 'Hand Over';
    if (state.winner !== null) title = `Player ${state.winner + 1} Wins!`;
    el('go-title').textContent = title;
    el('go-result').textContent = state.message;
    el('go-scores').textContent = `Scores — Player 1: ${state.scores[0]}  ·  Player 2: ${state.scores[1]}`;

    const handsEl = el('go-hands');
    handsEl.innerHTML = '';
    if (ki) {
        handsEl.appendChild(buildGoPlayer(ki.knocker, ki.knocker_melds, ki.knocker_deadwood, ki.knocker_dw_value, true));
        handsEl.appendChild(buildGoPlayer(1 - ki.knocker, ki.opponent_melds, ki.opponent_deadwood, ki.opponent_dw_value, false));
    }

    show('gameover-overlay');
}

function buildGoPlayer(idx, melds, deadwood, dwVal, isKnocker) {
    const div = document.createElement('div');

    const lbl = document.createElement('div');
    lbl.className = 'go-player-label';
    lbl.textContent = `Player ${idx + 1}${isKnocker ? ' (knocked)' : ''}  —  Deadwood: ${dwVal}`;
    div.appendChild(lbl);

    if (melds.length > 0) {
        const sl = document.createElement('div');
        sl.className = 'go-section-label';
        sl.textContent = 'Melds';
        div.appendChild(sl);
        melds.forEach(meld => {
            const row = document.createElement('div');
            row.className = 'go-row';
            meld.forEach(c => row.appendChild(makeCard(c, { inMeld: true, noHover: true })));
            div.appendChild(row);
        });
    }

    if (deadwood.length > 0) {
        const sl = document.createElement('div');
        sl.className = 'go-section-label';
        sl.textContent = 'Deadwood';
        div.appendChild(sl);
        const row = document.createElement('div');
        row.className = 'go-row';
        deadwood.forEach(c => row.appendChild(makeCard(c, { deadwood: true, noHover: true })));
        div.appendChild(row);
    }

    return div;
}

// ── Utils ─────────────────────────────────────────────────────────────────────
function el(id) { return document.getElementById(id); }
function show(id) { el(id).classList.remove('hidden'); }
function hide(id) { el(id).classList.add('hidden'); }

function showMsg(text, isError = false) {
    const m = el('message');
    m.textContent = text;
    m.className = 'message' + (isError ? ' error' : '');
    if (isError) setTimeout(() => { m.className = 'message'; m.textContent = state?.message ?? ''; }, 3000);
}

// ── Init ──────────────────────────────────────────────────────────────────────
async function init() {
    // Start straight into a game against the selected opponent (default: bot).
    await newGame();
}

init();

File: agent/game.py
from typing import List, Optional
import random

from agent.cards import Card, make_deck, is_set, is_run, deadwood_value, find_best_melds


class GameState:
    def __init__(self):
        self.scores = [0, 0]
        self.reset()

    def reset(self):
        deck = make_deck()
        random.shuffle(deck)
        self.hands: List[List[Card]] = [deck[:10], deck[10:20]]
        self.stock: List[Card] = deck[20:-1]
        self.discard_pile: List[Card] = [deck[-1]]
        self.current_player: int = 0
        self.phase: str = 'draw'
        self.drawn_card: Optional[Card] = None
        self.game_over: bool = False
        self.winner: Optional[int] = None
        self.knock_info: Optional[dict] = None
        self.turn_count: int = 0
        self.message: str = "Player 1's turn — draw a card"

    def draw(self, source: str) -> dict:
        if self.game_over:
            return {'ok': False, 'error': 'Game is over'}
        if self.phase != 'draw':
            return {'ok': False, 'error': 'Not time to draw'}

        p = self.current_player
        if source == 'discard':
            if not self.discard_pile:
                return {'ok': False, 'error': 'Discard pile is empty'}
            card = self.discard_pile.pop()
        elif source == 'stock':
            if not self.stock:
                if len(self.discard_pile) <= 1:
                    self._declare_draw()
                    return {'ok': True}
                top = self.discard_pile.pop()
                self.stock = self.discard_pile[:]
                random.shuffle(self.stock)
                self.discard_pile = [top]
            card = self.stock.pop()
        else:
            return {'ok': False, 'error': 'Invalid source'}

        self.hands[p].append(card)
        self.drawn_card = card
        self.phase = 'discard'
        src = 'the discard' if source == 'discard' else 'the stock'
        self.message = f"Player {p + 1} drew from {src} — select a card to discard or knock"
        return {'ok': True}

    def discard(self, card_str: str) -> dict:
        if self.game_over:
            return {'ok': False, 'error': 'Game is over'}
        if self.phase != 'discard':
            return {'ok': False, 'error': 'Draw a card first'}

        p = self.current_player
        card = self._find_in_hand(p, card_str)
        if card is None:
            return {'ok': False, 'error': f'{card_str} not in hand'}

        self.hands[p].remove(card)
        self.discard_pile.append(card)
        self.drawn_card = None
        self.turn_count += 1
        self.current_player = 1 - p
        self.phase = 'draw'
        self.message = f"Player {self.current_player + 1}'s turn — draw a card"
        return {'ok': True}

    def knock(self, card_str: str) -> dict:
        if self.game_over:
            return {'ok': False, 'error': 'Game is over'}
        if self.phase != 'discard':
            return {'ok': False, 'error': 'Draw a card first'}

        p = self.current_player
        card = self._find_in_hand(p, card_str)
        if card is None:
            return {'ok': False, 'error': f'{card_str} not in hand'}

        test_hand = [c for c in self.hands[p] if c is not card]
        _, dw_cards = find_best_melds(test_hand)
        dw = deadwood_value(dw_cards)
        if dw > 10:
            return {'ok': False, 'error': f'Deadwood is {dw} — need ≤ 10 to knock'}

        self.hands[p].remove(card)
        self.discard_pile.append(card)
        self.drawn_card = None
        self._resolve_knock(p)
        return {'ok': True}

    def _resolve_knock(self, knocker: int):
        opp = 1 - knocker
        k_melds, k_dw = find_best_melds(self.hands[knocker])
        k_dw_val = deadwood_value(k_dw)
        is_gin = k_dw_val == 0

        o_melds, o_dw = find_best_melds(self.hands[opp])

        if not is_gin:
            remaining = list(o_dw)
            changed = True
            while changed:
                changed = False
                for card in remaining[:]:
                    for meld in k_melds:
                        extended = meld + [card]
                        if is_set(extended) or is_run(extended):
                            meld.append(card)
                            remaining.remove(card)
                            changed = True
                            break
            o_dw = remaining

        o_dw_val = deadwood_value(o_dw)

        if is_gin:
            score = o_dw_val + 25
            self.scores[knocker] += score
            self.winner = knocker
            result = 'gin'
        elif o_dw_val <= k_dw_val:
            score = (k_dw_val - o_dw_val) + 25
            self.scores[opp] += score
            self.winner = opp
            result = 'undercut'
        else:
            score = o_dw_val - k_dw_val
            self.scores[knocker] += score
            self.winner = knocker
            result = 'knock'

        self.game_over = True
        self.knock_info = {
            'knocker': knocker,
            'knocker_melds': [[str(c) for c in m] for m in k_melds],
            'knocker_deadwood': [str(c) for c in k_dw],
            'knocker_dw_value': k_dw_val,
            'opponent_melds': [[str(c) for c in m] for m in o_melds],
            'opponent_deadwood': [str(c) for c in o_dw],
            'opponent_dw_value': o_dw_val,
            'result_type': result,
            'score': score,
        }

        w = f"Player {self.winner + 1}"
        if result == 'gin':
            self.message = f"GIN! {w} wins {score} points!"
        elif result == 'undercut':
            self.message = f"UNDERCUT! Player {opp + 1} undercuts — {w} wins {score} points!"
        else:
            self.message = f"Player {knocker + 1} knocks — {w} wins {score} points!"

    def _declare_draw(self):
        self.game_over = True
        self.winner = None
        self.message = "No cards left — hand is a draw!"

    def _find_in_hand(self, player: int, card_str: str) -> Optional[Card]:
        for c in self.hands[player]:
            if str(c) == card_str:
                return c
        return None

    def to_dict(self) -> dict:
        return {
            'current_player': self.current_player,
            'phase': self.phase,
            'hands': [[str(c) for c in h] for h in self.hands],
            'discard_top': str(self.discard_pile[-1]) if self.discard_pile else None,
            'discard_count': len(self.discard_pile),
            'stock_count': len(self.stock),
            'game_over': self.game_over,
            'winner': self.winner,
            'scores': self.scores,
            'message': self.message,
            'knock_info': self.knock_info,
            'turn_count': self.turn_count,
            'drawn_card': str(self.drawn_card) if self.drawn_card else None,
        }
